src/Container.php

```diff
@@ -739,29 +744,35 @@ public function getConfiguration(): Configuration
     }
 
     /**
-     * @param Closure|string|int|float|bool|object $value
+     * @param class-string<object> $id
+     * @param callable(self): object $value
      */
-    private function offsetAdd(string $id, $value): void
+    private function offsetAdd(string $id, callable $value): void
     {
         $this->keys[$id] = true;
 
-        if (is_callable($value)) {
-            $this->factories[$id] = $value;
-        } else {
-            $this->values[$id] = $value;
-        }
+        $this->factories[$id] = $value;
     }
 
-    private function get(string $id)
+    /**
+     * @template T of object
+     *
+     * @param class-string<T> $id
+     * @phpstan-return T
+     */
+    private function get(string $id): object
     {
         if (!isset($this->keys[$id])) {
             throw new InvalidArgumentException(sprintf('Unknown service "%s"', $id));
         }
 
         if (array_key_exists($id, $this->values)) {
-            return $this->values[$id];
+            $value = $this->values[$id];
+        } else {
+            $value = $this->values[$id] = $this->factories[$id]($this);
         }
+        assert($value instanceof  $id);
```

Comment:
weird CS here: also I don't get this assertion

Reply:
This is needed to understand that the `get` method returns an instance of the `id` class string. It means that phpstan knows this is correct.

```php
    public function getConfiguration(): Configuration
    {
        return $this->get(Configuration::class);
    }
```